### PR TITLE
Add cmp utility

### PR DESCRIPTION
**Pull request by kov**
The utility should support all the functionality supported by GNU cmp and perform quite a bit better. I split the commits for the actual utility in 3, with the base one being a full implementation of the features of GNU cmp, integration tests included.

The commits on top of that add optimizations which remove some and then all of rust fmt usage for the potentially long-running loop when --verbose is passed in. I wanted to separate them, first of all because it gives us good base for debugging potential issues with the optimized version, but secondly because I would understand the project preferring the slower version with more readable code.

That said, I do think it makes sense to adopt the optimized version as the gains are massive - on the order of 100% gains on my M1 Max comparing ~36M files that are completely different. Following tests run after warming up the I/O cache:

Baseline - GNU cmp

```
     > hyperfine --warmup 1 -i --output=pipe \
         'cmp -l huge huge.3'
     Benchmark 1: cmp -l huge huge.3
       Time (mean ± σ):      3.237 s ±  0.014 s    [User: 2.891 s, System: 0.341 s]
       Range (min … max):    3.221 s …  3.271 s    10 runs
    
       Warning: Ignoring non-zero exit code.
```

Unoptimized diffutils - ~74% of the time

```
     > hyperfine --warmup 1 -i --output=pipe \
         '../target/release/diffutils cmp -l huge huge.3'
     Benchmark 1: ../target/release/diffutils cmp -l huge huge.3
       Time (mean ± σ):      2.392 s ±  0.009 s    [User: 1.978 s, System: 0.406 s]
       Range (min … max):    2.378 s …  2.406 s    10 runs
    
       Warning: Ignoring non-zero exit code.
```

Optimized diffutils - 26% of the time

```
     > hyperfine --warmup 1 -i --output=pipe \
         '../target/release/diffutils cmp -l huge huge.3'
     Benchmark 1: ../target/release/diffutils cmp -l huge huge.3
       Time (mean ± σ):     849.5 ms ±   6.2 ms    [User: 538.3 ms, System: 306.8 ms]
       Range (min … max):   839.4 ms … 857.7 ms    10 runs
    
       Warning: Ignoring non-zero exit code.
```

### CODE REVIEW

**Comment by sylvestre** [comment]
could you please run your benchmarks with hyperfine? time is too limited

**Comment by kov** [comment]
> could you please run your benchmarks with hyperfine? time is too limited

Will do! Have begun working on fuzzing support, fwiw, need to do some refactoring as I currently use process::exit() as a shortcut in a lot of places.

**Comment by kov** [comment]
Added fuzz implementation to the first utility commit, ran it overnight with no issues spotted. Replaced my hand crafted benchmarks with hyperfine output (and made sure to have the system free of memory pressure / cpu interference).

I have tried to run the GNU tests, but the script fails here. I only changed it to also symlink diffutils to cmp and to run the cmp tests. Can anyone spot anything obvious before I start diving in to investigate?

```
kov@jabuticaba ~/P/diffutils (cmp)> ./tests/run-upstream-testsuite.sh
Fetching upstream test suite from https://git.savannah.gnu.org/git/diffutils.git
Running 31 tests
./tests/run-upstream-testsuite.sh: line 96: cd: gt-basic.*: No such file or directory
  basic                                    FAIL
./tests/run-upstream-testsuite.sh: line 96: cd: gt-bignum.*: No such file or directory
  bignum                                   FAIL
./tests/run-upstream-testsuite.sh: line 96: cd: gt-binary.*: No such file or directory
  binary                                   FAIL
./tests/run-upstream-testsuite.sh: line 96: cd: gt-brief-vs-stat-zero-kernel-lies.*: No such file or directory
  brief-vs-stat-zero-kernel-lies           FAIL
...
```

```diff
diff --git a/tests/run-upstream-testsuite.sh b/tests/run-upstream-testsuite.sh
index cb59834..f75b0b3 100755
--- a/tests/run-upstream-testsuite.sh
+++ b/tests/run-upstream-testsuite.sh
@@ -21,7 +21,7 @@
 # (e.g. 'dev' or 'test').
 # Unless overridden by the $TESTS environment variable, all tests in the test
 # suite will be run. Tests targeting a command that is not yet implemented
-# (e.g. cmp, diff3 or sdiff) are skipped.
+# (e.g. diff3 or sdiff) are skipped.
 
 scriptpath=$(dirname "$(readlink -f "$0")")
 rev=$(git rev-parse HEAD)
@@ -57,6 +57,7 @@ upstreamrev=$(git rev-parse HEAD)
 mkdir src
 cd src
 ln -s "$binary" diff
+ln -s "$binary" cmp
 cd ../tests
 
 if [[ -n "$TESTS" ]]
@@ -82,9 +83,9 @@ for test in $tests
 do
   result="FAIL"
   url="$urlroot$test?id=$upstreamrev"
-  # Run only the tests that invoke `diff`,
+  # Run only the tests that invoke `diff` or `cmp`,
   # because other binaries aren't implemented yet
-  if ! grep -E -s -q "(cmp|diff3|sdiff)" "$test"
+  if ! grep -E -s -q "(diff3|sdiff)" "$test"
   then
     sh "$test" 1> stdout.txt 2> stderr.txt && result="PASS" || exitcode=1
     json+="{\"test\":\"$test\",\"result\":\"$result\","
```

**Comment by oSoMoN** [comment]
clippy is reporting 9 trivial errors, could you address them to get the CI results green?

**Comment by oSoMoN** [comment]
> I have tried to run the GNU tests, but the script fails here. I only changed it to also symlink diffutils to cmp and to run the cmp tests. Can anyone spot anything obvious before I start diving in to investigate?

This is a regression that also affects main, so not introduced by your changes. I've filed #90 and am investigating it.

**Comment by oSoMoN** [comment]
> > I have tried to run the GNU tests, but the script fails here. I only changed it to also symlink diffutils to cmp and to run the cmp tests. Can anyone spot anything obvious before I start diving in to investigate?
> 
> This is a regression that also affects main, so not introduced by your changes. I've filed #90 and am investigating it.

With the fix in #91, I ran the test suite on your branch, and I'm seeing the following differences:
```diff
@@ -3,9 +3,9 @@
   basic                                    PASS
   bignum                                   PASS
   binary                                   FAIL
-  brief-vs-stat-zero-kernel-lies           SKIP
+  brief-vs-stat-zero-kernel-lies           FAIL
   bug-64316                                PASS
-  cmp                                      SKIP
+  cmp                                      FAIL
   colliding-file-names                     FAIL
   diff3                                    SKIP
   excess-slash                             FAIL
@@ -30,9 +30,9 @@
   strip-trailing-cr                        FAIL
   timezone                                 PASS
   colors                                   FAIL
-  y2038-vs-32bit                           SKIP
+  y2038-vs-32bit                           PASS
 
-Summary: TOTAL: 31 / PASS: 6 / FAIL: 20 / SKIP: 5
+Summary: TOTAL: 31 / PASS: 7 / FAIL: 22 / SKIP: 2
 
 Results written to /home/osomon/build/uutils/diffutils/tests/test-results.json
```

The failure in the `cmp` test needs to be investigated for sure, but this looks promising.

**Comment by kov** [comment]
Fixed the clippy complaints and quite close to getting the cmp test to pass locally. I am a bit confused, though, as my system's GNU cmp also doesn't pass the tests. I think the test infrastructure may not be sanitizing locale environments properly or there is something else on my Fedora it is not liking.

I got the failures on our tool down to a minimum, but still hit these, which the system GNU cmp also hits. Wonder if there is a weird corner case messing things up:

```
kov@jabuticaba ~/P/d/t/d/tests (master) [1]> cargo build; and env LANG=C ./cmp 
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
LC_ALL=C cmp -b bad bug
bad bug differ: byte 2, line 1 is 141 a 165 u
cmp: invalid --ignore-initial value '99999999999999999999999999999999999999999999999999999999999'
cmp: invalid --bytes value '99999999999999999999999999999999999999999999999999999999999'
```

**Comment by oSoMoN** [comment]
> I checked with strace that the cmp being run was appropriate fwiw. The last error message from GNU cmp is a real wth xD

I ran `LANG=C ./cmp` from the tests directory in a local checkout of the upstream repository, and I'm not seeing the "invalid --ignore-initial value" error that you're observing. I do see it when using cmp built from your branch.

Note that the `cmp` test script prepends `../src` to `$PATH`. Could it be that you had symlinked the wrong executable in there?
I'm seeing a different behaviour when I run that same test script against my system-wide `/usr/bin/cmp` installed from distro packages (using Ubuntu 24.04).

The following upstream commits seem relevant: [4ee8300](https://git.savannah.gnu.org/cgit/diffutils.git/commit?id=4ee8300) and [9c5fcbd](https://git.savannah.gnu.org/cgit/diffutils.git/commit?id=9c5fcbd).

**Comment by oSoMoN** [comment]
Comparing further the output of the `cmp` test script, it looks like your implementation conforms to the released version of GNU cmp (version 3.10), i.e. it chokes on large numbers passed to `-i`/`--ignore-initial`.

We have two options here:
 * align your implementation with the latest unreleased changes in the GNU diffutils repository
 * change `tests/run-upstream-testsuite.sh` to checkout a released version of the test suite, not the latest HEAD

I think that both approaches are equally interesting and valid (one favours the bleeding edge, the other compatibility with released versions), so feel free to pick whichever you prefer (I can help with updating the test script if you decide to go that way).

**Comment by kov** [comment]
Getting there, I went with updating the implementation to allow huge numbers - forward looking probably makes more sense =)

I just need to investigate this one now:

-  brief-vs-stat-zero-kernel-lies           SKIP
+  brief-vs-stat-zero-kernel-lies           FAIL

Will hopefully get to it this evening. Thanks for the help so far!

**Comment by oSoMoN** [comment]
> I just need to investigate this one now:
> 
>     * brief-vs-stat-zero-kernel-lies           SKIP
>     * brief-vs-stat-zero-kernel-lies           FAIL

I looked into it, and it turns out it's an issue in the test suite runner, not in your code. I filed #92 to track it.

**Comment by kov** [comment]
I created another fuzz target for parameter parsing and left it running overnight with no issues, do you think having this target is also useful? I think the hand-crafted argument parsing is more likely to go wrong, now that I think about it.

```rust
cat fuzz/fuzz_targets/fuzz_cmp_args.rs
#![no_main]
#[macro_use]
extern crate libfuzzer_sys;
use diffutilslib::cmp;

use std::ffi::OsString;

fuzz_target!(|x: Vec<OsString>| {
    let _ = cmp::parse_params(x.into_iter().peekable());
});
```

Any thoughts on how we could deny corpus entries to make it more useful? I thought about denying certain sizes if they do not have any -x or --whatever, or at least one of the known parameters. Like, if x.len() > 4 it needs to have some known parameters, but that excludes having over 4 positional parameters quite often, I suppose. Could be worth the tradeoff, still.

**Comment by oSoMoN** [comment]
> I created another fuzz target for parameter parsing and left it running overnight with no issues, do you think having this target is also useful? I think the hand-crafted argument parsing is more likely to go wrong, now that I think about it.

Yes, I think it is definitely useful. For the record, we implement argument parsing by hand because ready-made argument parsers like clap do not offer the flexibility we need to replicate the GNU diffutils arguments. We'd lose in compatibility what we would gain in code simplicity.

> Any thoughts on how we could deny corpus entries to make it more useful? I thought about denying certain sizes if they do not have any -x or --whatever, or at least one of the known parameters. Like, if x.len() > 4 it needs to have some known parameters, but that excludes having over 4 positional parameters quite often, I suppose. Could be worth the tradeoff, still.

I don't have prior experience with this, but perhaps we could use [a dictionary](https://llvm.org/docs/LibFuzzer.html#dictionaries)?

One additional thought: I see that you wrote integration tests for the new cmp command. Those are good and they provide a decent code coverage, but I would also suggest writing unit tests for `parse_params(…)` trying to bump the coverage as much as possible. With the combination of unit tests, integration tests and targeted fuzzing we should have our backs covered.

**Comment by kov** [comment]
Added a dictionary, but ended up going with some Corpus rejection as well, as using the dictionary requires passing parameters to cargo fuzz, may be forgotten. The github action looks like it would need to be split per target if we need to pass dictionaries there - which makes me realize I never added the fuzz target there, will fix right now.

In addition to the param parsing I added yesterday, I also added a few more integration tests to cover more of the fast paths that had no coverage (some of them were covered by the GNU cmp tests, but good to have our own tests as well I suppose).

**Comment by kov** [comment]
I may need to set up a build on a Windows VM xD

File: src/cmp.rs [hunk 1]
```diff
@@ -0,0 +1,1188 @@
...
+                    ));
+                }
+            };
+
+            let (n, overflows) = num.overflowing_mul(multiplier);
+            if overflows {
+                num = usize::MAX;
+            } else {
+                num = n;
+            }
```
Comment (oSoMoN) [inline src/cmp.rs:117]: With pattern matching this can be compacted to:
```rust
num = match num.overflowing_mul(multiplier) {
    (n, false) => n,
    _ => usize::MAX,
}
```

File: src/cmp.rs [hunk 2]
```diff
@@ -0,0 +1,1188 @@
...
+                )
+            } else {
+                let skip = parse_skip(&skip_desc, &skip_desc)?;
+                (skip, skip)
+            };
+            params.skip_a = Some(skip_a);
+            params.skip_b = Some(skip_b);
+            continue;
+        }
+        if param == "-s" || param == "-q" || param == "--quiet" || param == "--silent" {
```
Comment (oSoMoN) [inline src/cmp.rs:206]: `-q` doesn't appear to be a valid option for the GNU implementation. Is there a particular reason for accepting it here?

File: src/cmp.rs [hunk 3]
```diff
@@ -0,0 +1,1188 @@
...
+        return Err(format!(
+            "{}: options -l and -s are incompatible",
+            executable_str
+        ));
+    }
+
+    params.from = if let Some(from) = from {
+        from
+    } else if let Some(param) = opts.next() {
+        param
```
Comment (oSoMoN) [inline src/cmp.rs:249]: Isn't this already handled at the end of the while loop above?

File: src/cmp.rs [hunk 4]
```diff
@@ -0,0 +1,1188 @@
...
+    } else {
+        return Err(usage_string(&executable_str));
+    };
+    params.to = if let Some(to) = to {
+        to
+    } else if let Some(param) = opts.next() {
+        param
```
Comment (oSoMoN) [inline src/cmp.rs:256]: Same question as above.

File: src/cmp.rs [hunk 5]
```diff
@@ -0,0 +1,1188 @@
...
+    } else {
+        OsString::from("-")
+    };
+
+    // GNU cmp ignores positional skip arguments if -i is provided.
+    if params.skip_a.is_none() {
+        if let Some(skip_pos) = skip_pos1 {
+            params.skip_a = Some(skip_pos);
```
Comment (oSoMoN) [inline src/cmp.rs:264]: Could be slightly simplified to:
```rust
if skip_pos1.is_some() {
    params.skip_a = skip_pos1;
```

File: src/cmp.rs [hunk 6]
```diff
@@ -0,0 +1,1188 @@
...
+        } else if let Some(param) = opts.next() {
+            let param_str = param.to_string_lossy().to_string();
+            params.skip_a = Some(parse_skip(&param_str, &param_str)?);
+        }
```
Comment (oSoMoN) [inline src/cmp.rs:268]: Isn't this already handled at the end of the while loop above?

File: src/cmp.rs [hunk 7]
```diff
@@ -0,0 +1,1188 @@
...
+    quoted.push(byte);
+
+    // SAFETY: the checks and shifts we do above match what cat and GNU
+    // cmp do to ensure characters fall inside the ascii range.
+    unsafe { String::from_utf8_unchecked(quoted) }
+}
+
+// This function has been optimized to not use the Rust fmt system, which
+// leads to a massive speed up when processing large files: cuts the time
+// for comparing 2 ~36MB completely different files in half on an M1 Max.
```
Comment (oSoMoN) [inline src/cmp.rs:538]: I haven't run the benchmarks myself, I'll take your word for it. Out of curiosity, have you run the same benchmarks on medium-sized and small files?

File: tests/integration.rs [hunk 8]
```diff
@@ -299,3 +319,506 @@ fn compare_file_to_directory() -> Result<(), Box<dyn std::error::Error>> {
...
+    a.write_all(b"a\n").unwrap();
+
+    let b_path = tmp_dir.path().join("b");
+    let mut b = File::create(&b_path).unwrap();
+    b.write_all(b"a\n").unwrap();
+
+    let mut cmd = Command::cargo_bin("diffutils")?;
+    cmd.arg("cmp");
+    cmd.arg(&a_path).arg(&b_path);
+    cmd.assert().code(predicate::eq(0)).success();
```
[the remark this answers is not in the document]
Comment (kov) [reply tests/integration.rs:385]: I also added a bunch of checks for empty stderr, let me know if you think it's overkill. I suppose the more the merrier heh

File: src/main.rs [hunk 9]
```diff
@@ -3,119 +3,79 @@
...
-    let opts = env::args_os();
-    let params = parse_params(opts).unwrap_or_else(|error| {
-        eprintln!("{error}");
-        exit(2);
+    let mut args = std::env::args_os().peekable();
+
+    let exe_path = binary_path(&mut args);
+    let exe_name = name(&exe_path).unwrap_or_else(|| {
+        usage("<unknown binary>");
+        std::process::exit(1);
```
[the remark this answers is not in the document]
Comment (kov) [reply src/main.rs]: I ended up reworking this so that name() also panics. I think this is one of those cases in which there are more fundamental problems to worry about if these paths are hit, right?

[the remark this answers is not in the document]
Comment (oSoMoN) [reply src/main.rs]: Agreed.

File: src/cmp.rs (hunk 3 above)
Comment (oSoMoN) [reply src/cmp.rs:247]: Right, my mistake. My future self would appreciate a comment there because I know I might be caught again thinking that this is duplicated functionality. Given that this is already implemented like this in diff though, we can leave the addition of such a comment (in both places) for later.

File: src/cmp.rs [hunk 10]
```diff
@@ -0,0 +1,1188 @@
...
+            } else {
+                &params.to.to_string_lossy()
+            };
+            if params.verbose {
+                report_verbose_diffs(verbose_diffs, params)?;
+                report_eof(at_byte, at_line, start_of_line, eof_on, params);
+                return Ok(Cmp::Different);
+            } else {
+                report_eof(at_byte, at_line, start_of_line, eof_on, params);
+                return Ok(Cmp::Different);
```
[the remark this answers is not in the document]
Comment (oSoMoN) [reply src/cmp.rs]: You removed the `else` block, but I think you can go one step further and remove the call to `report_eof()` and the return statement from the `if` block.

[the remark this answers is not in the document]
Comment (kov) [reply src/cmp.rs]: Oh, indeed. I was blinded to this repetition while reworking this code when I initially wrote it, I guess. Thank you!

**Review by oSoMoN — APPROVED** [review]
This looks great now, thanks!
